powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthService.java

```diff
@@ -17,662 +17,1915 @@
+            throw ex;
+        } catch (Exception ex) {
+            logger.error("Unknown error occurred", ex);
+            throw new GenericServiceException(ServiceError.UNKNOWN_ERROR, ex.getMessage(), ex.getLocalizedMessage());
+        }
+    }
+
+    @Transactional
+    public UpdateStatusForActivationsResponse updateStatusForActivations(UpdateStatusForActivationsRequest request) throws GenericServiceException {
+        if (request.getActivationIds() == null || request.getActivationIds().isEmpty()) {
+            logger.warn("Invalid request parameter activationIds in method updateStatusForActivations");
+            // Rollback is not required, error occurs before writing to database
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_REQUEST);
+        }
+        try {
+            final List<String> activationIds = request.getActivationIds();
+            ActivationStatus activationStatus = null;
+            if (request.getActivationStatus() != null) {
+                activationStatus = activationStatusConverter.convert(request.getActivationStatus());
+            }
+            logger.info("UpdateStatusForActivationsRequest received");
+            final UpdateStatusForActivationsResponse response = behavior.getActivationServiceBehavior().updateStatusForActivation(activationIds, activationStatus);
+            logger.info("UpdateStatusForActivationsRequest succeeded");
+            return response;
+        } catch (RuntimeException | Error ex) {
+            logger.error("Runtime exception or error occurred, transaction will be rolled back", ex);
+            throw ex;
+        } catch (Exception ex) {
+            logger.error("Unknown error occurred", ex);
+            throw new GenericServiceException(ServiceError.UNKNOWN_ERROR, ex.getMessage(), ex.getLocalizedMessage());
+        }
+    }
+
+    @Transactional
+    public GetActivationStatusResponse getActivationStatus(GetActivationStatusRequest request) throws GenericServiceException {
+        if (request.getActivationId() == null) {
+            logger.warn("Invalid request parameter activationId in method getActivationStatus");
+            // Rollback is not required, database is not used for writing
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_REQUEST);
+        }
+        try {
+            final String activationId = request.getActivationId();
+            final String challenge = request.getChallenge();
+            logger.info("GetActivationStatusRequest received, activation ID: {}", activationId);
+            final GetActivationStatusResponse response = behavior.getActivationServiceBehavior().getActivationStatus(activationId, challenge, keyConvertor);
+            logger.info("GetActivationStatusResponse succeeded");
+            return response;
+        } catch (GenericServiceException ex) {
+            // already logged
+            throw ex;
+        } catch (RuntimeException | Error ex) {
```

Comment:
Catching `Error` is not a good practice, but let`s postpone it to another PR.
https://rules.sonarsource.com/java/RSPEC-1181

Reply:
OK, we can open a separate issue to it.